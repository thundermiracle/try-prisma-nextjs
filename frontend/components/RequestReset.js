import React from "react";
import PropTypes from "prop-types";

import { Mutation } from "react-apollo";
import gql from "graphql-tag";

import { CURRENT_USER_QUERY } from "./User";
import Form from "./styles/Form";
import ErrorMessage from "./ErrorMessage";

const REQUESTRESET_MUTATION = gql`
  mutation REQUESTRESET_MUTATION($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`;

const Requestreset = (props) => {
  const [state, setState] = React.useState({
    email: "",
  });

  const saveToState = (e) => {
    setState({
      ...state,
      [e.target.name]: e.target.value,
    });
  };

  return (
    <Mutation mutation={REQUESTRESET_MUTATION} variables={state}>
      {(requestreset, { error, loading, called }) => {
        return (
          <Form
            method="post"
            onSubmit={(e) => {
              e.preventDefault();
              requestreset();
              setState({ email: "" });
            }}
          >
            <fieldset disabled={loading} aria-busy={loading}>
              <h2>Request Password Reset</h2>
              <ErrorMessage error={error} />
              {!error && !loading && called && (
                <p>Success! Check your email for a reset link!</p>
              )}
              <label htmlFor="email">
                Email
                <input
                  type="text"
                  name="email"
                  placeholder="email"
                  value={state.email}
                  onChange={saveToState}
                />
              </label>

              <button type="submit">Request!</button>
            </fieldset>
          </Form>
        );
      }}
    </Mutation>
  );
};

Requestreset.propTypes = {};

export default Requestreset;
