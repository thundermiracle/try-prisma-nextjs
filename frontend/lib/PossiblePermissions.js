const PossiblePermissions = [
  "ADMIN",
  "USER",
  "ITEMCREATE",
  "ITEMUPDATE",
  "ITEMDELETE",
  "PERMISISIONUPDATE",
];

export default PossiblePermissions;
