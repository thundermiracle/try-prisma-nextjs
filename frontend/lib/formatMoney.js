export default function (amount) {
  const options = { style: "currency", currency: "JPY" };

  const formatter = new Intl.NumberFormat("ja-JP", options);
  return formatter.format(amount);
}
