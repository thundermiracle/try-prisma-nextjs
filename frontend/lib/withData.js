import withApollo from "next-with-apollo";
import ApolloClient from "apollo-boost";
import { endpoint, prodEndpoint } from "../config";

import { LOCAL_STATE_QUERY } from "../components/Cart";

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === "development" ? endpoint : prodEndpoint,
    request: (operation) => {
      operation.setContext({
        fetchOptions: {
          credentials: "include",
        },
        headers,
      });
    },
    // client local state
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, args, { cache }) {
            // read the cartOpen from cache
            const cachedData = cache.readQuery({ query: LOCAL_STATE_QUERY });

            // toggle cartOpen state
            const data = {
              data: { ...cachedData, cartOpen: !cachedData.cartOpen },
            };
            cache.writeData(data);

            // return
            return data;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo(createClient);
