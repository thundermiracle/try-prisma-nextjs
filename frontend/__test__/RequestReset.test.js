import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import { MockedProvider } from "@apollo/react-testing";

import { GraphQLError } from "graphql";
import { waitForComponentToPaint } from "../lib/testUtils";
import RequestReset, {
  REQUESTRESET_MUTATION,
} from "../components/RequestReset";

const CorrectEmail = "daniel@thundermiracle.com";
const IncorrectEmail = "test@thundermiracle.com";

const mockSuccess = [
  {
    request: {
      query: REQUESTRESET_MUTATION,
      variables: { email: CorrectEmail },
    },
    result: {
      data: {
        requestReset: {
          message: "success",
          __typename: "message",
        },
      },
    },
  },
];

const mockFailed = [
  {
    request: {
      query: REQUESTRESET_MUTATION,
      variables: { email: IncorrectEmail },
    },
    result: {
      data: {
        requestReset: {
          message: "failed",
          __typename: "message",
        },
      },
      errors: [new GraphQLError("GraphQL Error!")],
    },
  },
];

describe("<RequestReset />", () => {
  it("render properly", () => {
    const wrapper = mount(
      <MockedProvider mocks={mockSuccess}>
        <RequestReset />
      </MockedProvider>,
    );

    expect(toJSON(wrapper.find("Form"))).toMatchSnapshot();
  });

  it("success", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mockSuccess}>
        <RequestReset />
      </MockedProvider>,
    );

    // simulate inputting email
    wrapper.find("input").simulate("change", {
      target: { name: "email", value: CorrectEmail },
    });
    wrapper.find("form").simulate("submit");

    await waitForComponentToPaint(wrapper);

    expect(wrapper.find("fieldset p").text()).toContain("Check your email");
  });

  it("failed", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mockFailed}>
        <RequestReset />
      </MockedProvider>,
    );

    // simulate inputting email
    wrapper.find("input").simulate("change", {
      target: { name: "email", value: IncorrectEmail },
    });
    wrapper.find("form").simulate("submit");

    await waitForComponentToPaint(wrapper);

    expect(wrapper.find("DisplayError").text()).toContain("GraphQL Error!");
  });
});
