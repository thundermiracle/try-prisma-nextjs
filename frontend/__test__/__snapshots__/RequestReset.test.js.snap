// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset /> render properly 1`] = `
<Form
  method="post"
  onSubmit={[Function]}
>
  <form
    className="Form-sc-1xszr8q-0 krRxky"
    method="post"
    onSubmit={[Function]}
  >
    <fieldset
      aria-busy={false}
      disabled={false}
    >
      <h2>
        Request Password Reset
      </h2>
      <DisplayError
        error={Object {}}
      />
      <label
        htmlFor="email"
      >
        Email
        <input
          name="email"
          onChange={[Function]}
          placeholder="email"
          type="text"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Request!
      </button>
    </fieldset>
  </form>
</Form>
`;
